EasyLM/models/llama/llama_train_online_dpo.py

```diff
@@ -0,0 +1,544 @@
+            policy_params['params'], reference_params, reward_params,
+            policy_model, reference_model, reward_model,
+            rng_generator(), batch, tokenizer,
+        )
+        grad_fn = jax.value_and_grad(loss_and_metrics, has_aux=True)
+        (_, metrics), grads = grad_fn(policy_train_state.params)
+        policy_train_state = policy_train_state.apply_gradients(grads=grads)
+        return policy_train_state, rng_generator(), batch, metrics
+    
+    sharded_dpo_forward_backward = pjit(
+        dpo_forward_backward_wrapper,
+        in_shardings=(train_state_partition_policy, train_state_partition_policy.params, train_state_partition_reward.params, PS(), PS()),
+        out_shardings=(train_state_partition_policy, PS(), PS(), PS()),
+        donate_argnums=(0, 3),  # policy train state and rng
+    )
+
+    checkpointer = StreamingCheckpointer(
+        FLAGS.checkpointer, logger.output_dir,
+        enable=jax.process_index() == 0,
+    )
+    def save_checkpoint(policy_train_state, step, milestone=False):
+        metadata = dict(
+            step=step,
+            variant=variant,
+            flags=flags_config_dict,
+            llama_config=llama_config_policy.to_dict(),
+        )
+        checkpointer.save_all(
+            train_state=policy_train_state,
+            gather_fns=gather_fns_policy,
+            metadata=metadata,
+            milestone=milestone,
+            step=step,
+        )
+
+    mesh = LLaMAConfig.get_jax_mesh(FLAGS.mesh_dim)
+    with mesh:
+        # Load policy
+        policy_train_state, policy_params = None, None
+        if FLAGS.load_checkpoint_policy != '':
+            print("Loading checkpoint (policy) ... (may take time to download)")
+            policy_train_state, policy_params = checkpointer.load_trainstate_checkpoint(FLAGS.load_checkpoint_policy, train_state_shapes_policy, shard_fns_policy)
+            print("Checkpoint (policy) loaded.")
+        if policy_train_state is None:
+            if policy_params is None:
+                policy_train_state = sharded_init_fn_policy(next_rng())
+            else:
+                if not FLAGS.use_tpu:
+                    policy_params = flax.core.frozen_dict.unfreeze(policy_params)
+                policy_train_state = sharded_create_trainstate_from_params_policy(policy_params)
+                del policy_params
+
+        # Load reference
+        reference_params = None
+        if FLAGS.load_checkpoint_policy != '':
+            print("Loading checkpoint (reference) ... (may take time to download)")
+            reference_train_state, reference_params = checkpointer.load_trainstate_checkpoint(FLAGS.load_checkpoint_policy, train_state_shapes_policy, shard_fns_policy)
+            print("Checkpoint (reference) loaded.")
+            assert reference_train_state is None
+        if reference_params is None:
+            reference_params = copy.deepcopy(policy_train_state.params)
```

Comment:
You can probably just copy these params, w/o loading the ckpt again from disk.

Reply:
Does that 'just work'? I was a bit worried about deepcopy's compatability with sharded data.

Reply:
Oh okay actually i haven't tested this. Just realized that my code path never hit the deepcopy thing. Ignore this comment :)